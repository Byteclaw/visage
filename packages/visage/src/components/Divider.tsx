import {
  ExtractVisageComponentProps,
  markAsVisageComponent,
  VisageComponent,
} from '@byteclaw/visage-core';
import React, { forwardRef } from 'react';
import { createComponent } from '../core';
import { booleanVariant } from '../variants';
import { Text } from './Text';

const DividerLine = createComponent('div', {
  displayName: 'DividerLine',
  styles: props => ({
    borderWidth: 0,
    display: 'block',
    borderColor: 'currentColor',
    borderStyle: 'solid',
    mx: 0,
    my: 0,
    ...(props.vertical
      ? { borderLeftWidth: '1px', height: '100%', minWidth: '1px' }
      : { borderBottomWidth: '1px', minHeight: '1px', width: '100%' }),
  }),
  variants: [booleanVariant('vertical', true)],
});

const DividerBase = createComponent('div', {
  displayName: 'Divider',
  styles: props => ({
    display: 'flex',
    alignItems: 'center',
    width: '100%',
    ...(props.vertical
      ? { flexDirection: 'column', height: '100%', maxWidth: '100%' }
      : { width: '100%' }),
  }),
  variants: [booleanVariant('vertical', true)],
});

const DividerLabel = createComponent(Text, {
  displayName: 'DividerLabel',
  styles: props => (props.vertical ? { py: 1 } : { px: 1 }),
  variants: [booleanVariant('vertical', true)],
});

interface DividerProps extends ExtractVisageComponentProps<typeof DividerBase> {
  label?: string;
  labelProps?: ExtractVisageComponentProps<typeof DividerLine>;
  lineProps?: ExtractVisageComponentProps<typeof DividerLine>;
  vertical?: boolean;
}

export const Divider: VisageComponent<DividerProps> = React.memo(
  forwardRef(
    (
      { label, labelProps, lineProps, vertical, ...restProps }: DividerProps,
      ref,
    ) => {
      // if label is provided, split horizontal line to 2 elements
      if (!label) {
        return (
          <DividerBase
            {...restProps}
            ref={ref as any}
            role="separator"
            vertical={vertical}
          >
            <DividerLine {...lineProps} vertical={vertical} />
          </DividerBase>
        );
      }

      return (
        <DividerBase
          {...restProps}
          aria-label={label}
          role="separator"
          ref={ref as any}
          vertical={vertical}
        >
          <DividerLine {...lineProps} vertical={vertical} />
          <DividerLabel {...labelProps} role="presentation" vertical={vertical}>
            {label}
          </DividerLabel>
          <DividerLine {...lineProps} vertical={vertical} />
        </DividerBase>
      );
    },
  ),
) as any;

markAsVisageComponent(Divider);
