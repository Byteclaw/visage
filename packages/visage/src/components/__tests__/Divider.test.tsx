import React from 'react';
import { Divider } from '../Divider';
import { createTestTheme, render } from './render';

describe('Divider', () => {
  it('renders correctly (horizontal - default)', () => {
    const { asFragment } = render(<Divider />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with a label', () => {
    const { asFragment } = render(<Divider label="OR" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly as vertical', () => {
    const { asFragment } = render(<Divider vertical />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly as vertical with a label', () => {
    const { asFragment } = render(<Divider label="OR" vertical />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by Divider face', () => {
    const { asFragment } = render(<Divider label="OR" />, {
      ds: {
        theme: createTestTheme({
          Divider: {
            color: 'black',
          },
        }),
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by DividerLine face', () => {
    const { asFragment } = render(<Divider label="OR" />, {
      ds: {
        theme: createTestTheme({
          DividerLine: {
            color: 'black',
          },
        }),
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by DividerLabel face', () => {
    const { asFragment } = render(<Divider label="OR" />, {
      ds: {
        theme: createTestTheme({
          DividerLabel: {
            color: 'black',
          },
        }),
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
