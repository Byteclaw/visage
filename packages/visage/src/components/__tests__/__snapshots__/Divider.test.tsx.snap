// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider is extendable by Divider face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-bottom-width: 1px;
  min-height: 1px;
  width: 100%;
}

.emotion-1 {
  display: inline;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  padding: 0px;
  margin: 0px;
  margin-bottom: 0px;
  padding-left: 8px;
  padding-right: 8px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: rgb(0,0,0);
}

<div
    aria-label="OR"
    class="emotion-3"
    data-vertical="false"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="false"
    />
    <span
      class="emotion-1"
      data-vertical="false"
      role="presentation"
    >
      OR
    </span>
    <div
      class="emotion-0"
      data-vertical="false"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider is extendable by DividerLabel face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-bottom-width: 1px;
  min-height: 1px;
  width: 100%;
}

.emotion-1 {
  display: inline;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  padding: 0px;
  margin: 0px;
  margin-bottom: 0px;
  padding-left: 8px;
  padding-right: 8px;
  color: rgb(0,0,0);
}

<div
    aria-label="OR"
    class="emotion-3"
    data-vertical="false"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="false"
    />
    <span
      class="emotion-1"
      data-vertical="false"
      role="presentation"
    >
      OR
    </span>
    <div
      class="emotion-0"
      data-vertical="false"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider is extendable by DividerLine face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-1 {
  display: inline;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  padding: 0px;
  margin: 0px;
  margin-bottom: 0px;
  padding-left: 8px;
  padding-right: 8px;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-bottom-width: 1px;
  min-height: 1px;
  width: 100%;
  color: rgb(0,0,0);
}

<div
    aria-label="OR"
    class="emotion-3"
    data-vertical="false"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="false"
    />
    <span
      class="emotion-1"
      data-vertical="false"
      role="presentation"
    >
      OR
    </span>
    <div
      class="emotion-0"
      data-vertical="false"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider renders correctly (horizontal - default) 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-bottom-width: 1px;
  min-height: 1px;
  width: 100%;
}

<div
    class="emotion-1"
    data-vertical="false"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="false"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider renders correctly as vertical 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  max-width: 100%;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-left-width: 1px;
  height: 100%;
  min-width: 1px;
}

<div
    class="emotion-1"
    data-vertical="true"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="true"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider renders correctly as vertical with a label 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  max-width: 100%;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-left-width: 1px;
  height: 100%;
  min-width: 1px;
}

.emotion-1 {
  display: inline;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  padding: 0px;
  margin: 0px;
  margin-bottom: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
}

<div
    aria-label="OR"
    class="emotion-3"
    data-vertical="true"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="true"
    />
    <span
      class="emotion-1"
      data-vertical="true"
      role="presentation"
    >
      OR
    </span>
    <div
      class="emotion-0"
      data-vertical="true"
    />
  </div>
</DocumentFragment>
`;

exports[`Divider renders correctly with a label 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-0 {
  border-width: 0;
  display: block;
  border-color: rgb(184,184,184);
  border-style: solid;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  border-bottom-width: 1px;
  min-height: 1px;
  width: 100%;
}

.emotion-1 {
  display: inline;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit;
  padding: 0px;
  margin: 0px;
  margin-bottom: 0px;
  padding-left: 8px;
  padding-right: 8px;
}

<div
    aria-label="OR"
    class="emotion-3"
    data-vertical="false"
    role="separator"
  >
    <div
      class="emotion-0"
      data-vertical="false"
    />
    <span
      class="emotion-1"
      data-vertical="false"
      role="presentation"
    >
      OR
    </span>
    <div
      class="emotion-0"
      data-vertical="false"
    />
  </div>
</DocumentFragment>
`;
